### ISSUE TITLE
Discussion: use materialized view to maintain table statistics

### ISSUE THREAD

**Issue by chenzl25** [risingwavelabs/risingwave#5529]
### Is your feature request related to a problem? Please describe.

Some table statistics is useful to query optimization. Such as:
- table row count
- column NDV (number  of  distinct value)
- column null ratio
- column histogram
- column TopN (top frequency value)

Because statistics can be estimated approximately instead of exactly. With our stream system, maybe we can maintain our statistics incrementally with low overhead? Feel free to share your idea, just a discussion here.

**Comment by BowenXiao1999** [comment]
What's your plan for collect this? 

Table statistics should be in meta, data are flowing through CN, maybe we need to have some stat reporters in CN?

**Comment by chenzl25** [comment]
> What's your plan for collect this?
> 
> Table statistics should be in meta, data are flowing through CN, maybe we need to have some stat reporters in CN?

I think we can use a background task (as you say a reporter) to periodically collect statistics based on our internal stat mv and send them to the meta. But the more important thing is how to maintain statistics increasingly and efficiently.

**Comment by jon-chuang** [comment]
Could this be something we implement at state store layer? The data we require seems quite generic.

**Comment by chenzl25** [comment]
> Could this be something we implement at state store layer? The data we require seems quite generic.
> 
> In that case, wrt our discussion about insert + update v.s. upsert (#3893), perhaps insert + update can give a better idea about row count without needing to do a read.
> 
> Or, we can ask the operator layer to provide a hint about whether an upsert is an insert or an update. Thoughts? @BugenZhao @richardchien

State store is lack of schema, but for a table we need column level statistics.

**Comment by chenzl25** [comment]
For append only table, there exists some efficient incremental methods to maintain statistics.
- Table row count: we can use a simple counter  is enough.
- Column NDV: we can use HyperLogLog Sketch to estimates the cardinality with limited memory.
- null ratio, Histogram and TopN: we can keep sampling const number rows (like 100,000 rows at most) by reservoir sampling and calculate null ratio, Histogram and TopN based on the sampling data periodically.

While for table with update/delete, we need to do some investigation.

**Comment by jon-chuang** [comment]
> State store is lack of schema, but for a table we need column level statistics.

Apologies, I meant `StateTable`. I guess you are referring to collecting at StateTable level. Correct me if I am wrong.

For hyperloglog, for stream, we actually need something slightly more complicated. We already have it implemented. https://github.com/risingwavelabs/risingwave/issues/2727

**Comment by chenzl25** [comment]
```sql
create materialized view sbtest1_statistic as 
select count(*) as table_count, 
approx_count_distinct(id) id_ndv,  
approx_count_distinct(k) k_ndv,  
approx_count_distinct(c) c_ndv 
from sbtest1;
```

**Comment by chenzl25** [comment]
> > State store is lack of schema, but for a table we need column level statistics.
> 
> Apologies, I meant `StateTable`. I guess you are referring to collecting at StateTable level? Correct me if I am wrong.
> 
> What do you mean to "use materialized view"? Do you just mean to maintain the statistics via a stateful and incremenetal approach?
> 
> For hyperloglog, for stream, we actually need something slightly more complicated. We already have it implemented. #2727

Cool, happy to know that our hyperloglog already support update/delete.

**Comment by lmatz** [comment]
> They can offload this kind of workload to our system if we can handle it efficiently.

Good point! And my experience is that CBO often delivers horrible plans without accurate statistics. 
Potentially a big win for RisingWave.

**Comment by jon-chuang** [comment]
> Here is an example. Just use a sql to express the statistics what we need.

Cool. I guess we just need to not collect statistics on this "system" statistics table, as otherwise there will be infinite tables :)

**Comment by jon-chuang** [comment]
I see, I guess by `Table` we are referring either to source or MV (and not operator `StateTable`), and therefore by statistics, we are referring to MV on MV.

Let me think whether we can collect other statistics at operator level that are potentially useful for CBO in our system, such as predicate selectivity. Maybe we can combine these two to get better CBO.

**Comment by st1page** [comment]
I think there are two issues: 
- I check the current stream `approx_count_distinct` implementation and it might have some correctness issues. https://github.com/risingwavelabs/risingwave/issues/5616 
- IIUC, the optimizer does not need very fresh statistics. And better freshness means higher cost in our system due to the higher frequencies of barriers and checkpoints. So maybe we can not accept the statistics MV and users MV have the same freshness. We used to discuss some designs to maintain different freshness levels for different MV in our system which is too complex for us. I am not sure if the statistics a good use case for it. @fuyufjh @hzxa21 @BugenZhao

**Comment by fuyufjh** [comment]
In my mind the problem is on the use cases of statistics. Eventually we will have 3 kinds of database objects:

- source (not selectable, but can be used to create MV)
- table
- materialized view

As we know, for streaming queries, it may be created on source and/or newly-created tables, so statistics could be insufficient for optimizer.

As for batch queries, currently we don't encourage users to run complex queries in RisingWave. Instead, we encourage creating MV for that.

**Comment by chenzl25** [comment]
I think we should hold on until we figure out that the benefits of introducing statistics can totally outweigh its costs.
